libs/SmartStore/SmartStore/Classes/SFSmartStore.m

```diff
@@ -810,14 +820,28 @@ - (id)loadExternalSoupEntry:(NSNumber *)soupEntryId
                                                     error:&error];
     [inputStream close];
     if (!result && error) {
-        NSString *errorMessage = [NSString stringWithFormat:@"Loading external soup from file failed! encrypted: %@, soupEntryId: %@, soupTableName: %@, filePath: '%@', error: %@.",
-                                  keyBlock ? @"YES" : @"NO",
-                                  soupEntryId,
-                                  soupTableName,
-                                  filePath,
-                                  error];
-        NSAssert(NO, errorMessage);
-        [SFSDKSmartStoreLogger e:[self class] format:errorMessage];
+        
+        // It is possible that file was previously encrypted with nill IV. For Backward
+        // compatability reasons we will attempt to load the file with a nill IV. If itmanages
+        // to load one  it will encryptthe file with anIV and store it back.
+        if (initializationVector) {
+            result = [self loadExternalSoupEntry:soupEntryId soupTableName:soupTableName withKey:key andIV:nil];
+            //lets write it back.
+            BOOL migrated = [self saveSoupEntryExternally:result soupEntryId:soupEntryId soupTableName:soupTableName];
+            if (!migrated) {
+                [SFSDKSmartStoreLogger e:[self class] format:@"Attempt to migrate an encrypted externally saved soup '%@' with a null IV  failed.", soupTableName];
+            }
+            
+        } else {
+           NSString *errorMessage = [NSString stringWithFormat:@"Loading external soup from file failed! encrypted: %@, soupEntryId: %@, soupTableName: %@, filePath: '%@', error: %@.",
+                                      key ? @"YES" : @"NO",
+                                      soupEntryId,
+                                      soupTableName,
+                                      filePath,
+                                      error];
+            NSAssert(NO, errorMessage);
```

Comment:
I know this was here before, but this is not a good idea as a runtime exception generator, and it's exacerbated in this PR by further hiding the source of the issue—i.e. it doesn't fail fast.  The reason it's not good as a runtime check is because, if assertions are turned off, it's not going to fail _at all_.  Instead, it's going to drop through, and you're going to end up with unexpected behavior that's hard to track down.

`NSAssert` is good for sanity checks on arguments that would reasonably be caught during development cycles.  It's not good for user input, or data that's otherwise dynamically populated as a runtime experience.  If you truly need to fail the app in this scenario, you should raise an exception.

Reply:
@khawkins I will replace it with an exception.